internal/buildengine/plugin.go

```diff
@@ -0,0 +1,351 @@
+func (e AutoRebuildStartedEvent) ModuleName() string { return e.Module }
+
+// AutoRebuildEndedEvent is sent when the plugin ends an automatic rebuild.
+type AutoRebuildEndedEvent struct {
+	Module string
+	Result either.Either[BuildResult, error]
+}
+
+func (AutoRebuildEndedEvent) pluginEvent()         {}
+func (e AutoRebuildEndedEvent) ModuleName() string { return e.Module }
+
+// LanguagePlugin handles building and scaffolding modules in a specific language.
+type LanguagePlugin interface {
+	// Topic for all update events from the plugin
+	// The same topic must be returned each time this method is called
+	Updates() *pubsub.Topic[PluginEvent]
+
+	// CreateModule creates a new module in the given directory with the given name and language.
+	// Replacements and groups are special cases until plugins can provide their parameters.
+	CreateModule(ctx context.Context, config moduleconfig.ModuleConfig, includeBinDir bool, replacements map[string]string, group string) error
+
+	// GetDependencies returns the dependencies of the module.
+	GetDependencies(ctx context.Context) ([]string, error)
+
+	// Build builds the module with the latest config and schema.
+	// In dev mode, plugin is responsible for automatically rebuilding as relevant files within the module change,
+	// and publishing these automatic builds updates to Updates().
+	Build(ctx context.Context, projectRoot string, config moduleconfig.ModuleConfig, sch *schema.Schema, buildEnv []string, devMode bool) (BuildResult, error)
+
+	// Kill stops the plugin and cleans up any resources.
+	Kill(ctx context.Context) error
+}
+
+// PluginFromConfig creates a new language plugin from the given config.
+func PluginFromConfig(ctx context.Context, config moduleconfig.ModuleConfig, projectRoot string) (p LanguagePlugin, err error) {
+	switch config.Language {
+	case "go":
+		return newGoPlugin(ctx, config), nil
+	case "java", "kotlin":
+		return newJavaPlugin(ctx, config), nil
+	case "rust":
+		return newRustPlugin(ctx, config), nil
+	default:
+		return p, fmt.Errorf("unknown language %q", config.Language)
+	}
+}
+
+var scaffoldFuncs = template.FuncMap{
+	"snake":          strcase.ToLowerSnake,
+	"screamingSnake": strcase.ToUpperSnake,
+	"camel":          strcase.ToUpperCamel,
+	"lowerCamel":     strcase.ToLowerCamel,
+	"strippedCamel":  strcase.ToUpperStrippedCamel,
+	"kebab":          strcase.ToLowerKebab,
+	"screamingKebab": strcase.ToUpperKebab,
+	"upper":          strings.ToUpper,
+	"lower":          strings.ToLower,
+	"title":          strings.Title,
+	"typename":       schema.TypeName,
+}
+
+//sumtype:decl
+type pluginCommand interface {
+	pluginCmd()
+}
+
+type buildCommand struct {
+	projectRoot string
+	config      moduleconfig.ModuleConfig
+	schema      *schema.Schema
+	buildEnv    []string
+	devMode     bool
+
+	result chan either.Either[BuildResult, error]
+}
+
+func (buildCommand) pluginCmd() {}
+
+type dependenciesFunc = func() ([]string, error)
```

Comment:
Typically I would suggest only using `=` in very specific situations, it's _generally_ (but not always) more likely you want to use `type foo bar`